src/data_info/services/groups.clj

```diff
@@ -0,0 +1,77 @@
+(ns data-info.services.groups
+  (:require [clojure.tools.logging :as log]
+            [slingshot.slingshot :refer [throw+]]
+            [clj-jargon.users :as users]
+            [clojure.string :as string]
+            [data-info.services.users :as di-users]
+            [data-info.util.config :as cfg]
+            [data-info.util.irods :as irods]
+            [data-info.util.validators :as validators]))
+
+(defn get-group
+  [{user :user} group-name]
+  (irods/with-jargon-exceptions [cm]
+    (validators/user-exists cm user)
+    (validators/group-exists cm group-name)
+    {:group-name group-name :members (users/list-group-members cm group-name)}))
+
+(defn create-group
+  [{user :user} {group-name :group-name members :members}]
+  (irods/with-jargon-exceptions [cm]
+    (validators/user-exists cm user)
+    (validators/user-is-group-admin cm user)
+    (validators/group-does-not-exist cm group-name)
+    (validators/all-users-exist cm members)
+    (users/create-user-group cm group-name)
+    (when (seq members)
+      (dorun (map
+               (partial users/add-to-group cm group-name)
+               members)))
+    {:group-name group-name :members (users/list-group-members cm group-name)}))
+
+(defn update-group-members
+  [{user :user :as params} {members :members} group-name]
+  (irods/with-jargon-exceptions [cm]
+    (validators/user-exists cm user)
+    (validators/user-is-group-admin cm user)
+    (validators/group-exists cm group-name)
+    (validators/all-users-exist cm members)
+    (let [current-members (users/list-group-members cm group-name)]
+      (loop [old-members (sort current-members)
+             new-members (sort members)]
+        ;; compare the head of both lists and determine if anything needs adding or removing
+        (let [c (first old-members)
+              n (first new-members)]
+          (cond
```

Comment:
Is it possible to use `clojure.set` functions for this?

Maybe you'd have to map `di-users/qualify-username` to the new members first, then add them and the old members to their own sets, then `clojure.set/difference` one from the other to see what's left for adding or removing from the group?

Reply:
hm, I guess that might be cleaner, and these are just strings after all. I was a bit worried about groups with many members causing issues, but it also occurs to me that since I'm sorting them anyway it probably doesn't make a difference in terms of anything like memory. I might be able to `reduce` directly into a set and apply `qualify-username` when I do, and then unqualify the usernames after doing set differences and such, then pass to clj-jargon